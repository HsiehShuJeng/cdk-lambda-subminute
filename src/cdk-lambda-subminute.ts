// import * as sfn from '@aws-cdk/aws-stepfunctions';
// import * as tasks from '@aws-cdk/aws-stepfunctions-tasks';
import * as path from 'path';
import events = require('@aws-cdk/aws-events');
import targets = require('@aws-cdk/aws-events-targets');
import * as iam from '@aws-cdk/aws-iam';

import { IFunction, Runtime, Tracing } from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { RetentionDays } from '@aws-cdk/aws-logs';
import * as sfn from '@aws-cdk/aws-stepfunctions';
import * as tasks from '@aws-cdk/aws-stepfunctions-tasks';
import * as cdk from '@aws-cdk/core';

export interface LambdaSubminuteProps {
  /**
   * The Lambda function that is going to be executed per time unit less than one minute.
   */
  readonly targetFunction: IFunction;
  /**
   * A pattern you want this statemachine to be executed.
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html
   *
   * @default cron(50/1 15-17 ? * * *) UTC+0 being run every minute starting from 15:00 PM to 17:00 PM.
   */
  readonly conjobExpression?: string;
}

export class LambdaSubminute extends cdk.Construct {
  /**
   * The Lambda function that plays the role of the iterator.
   */
  readonly iteratorFunction: IFunction;
  /**
   * The ARN of the state machine that executes the target Lambda function per time unit less than one minute.
   */
  readonly stateMachineArn: string;
  constructor(parent: cdk.Construct, name: string, props: LambdaSubminuteProps) {
    super(parent, name);
    const iterator = new IteratorLambda(this, 'IteratorLambda', { targetFunction: props.targetFunction });
    this.iteratorFunction = iterator.function;
    const subminuteStateMachine = new SubminuteStateMachine(this, 'SubminuteStateMachine', {
      stateMachineName: 'lambda-subminute-statemachine',
      targetFunction: props.targetFunction,
      iteratorFunction: this.iteratorFunction,
    });
    this.stateMachineArn = subminuteStateMachine.stateMachine.stateMachineArn;

    const startRule = new events.Rule(this, 'StartSubminuteStateMachine', {
      schedule: events.Schedule.expression(props.conjobExpression ?? 'cron(50/1 15-17 ? * * *)'),
      ruleName: 'subminute-statemachine-lambda-rule',
      description: `A rule to run the subminute state machine, i.e. ${subminuteStateMachine.stateMachine.stateMachineName}`,
    });
    startRule.addTarget(new targets.SfnStateMachine(
      subminuteStateMachine.stateMachine, {
        input: events.RuleTargetInput.fromObject({
          iterator: {
            index: 0,
            count: 6,
          },
        }),
      }));
  }
}

interface IteratorLambdaProps {
  /**
     * The Lambda function that is going to be executed per time unit less than one minute.
     */
  targetFunction: IFunction;
}

class IteratorLambda extends cdk.Construct {
  /**
     * A Lambda function that plays the role of the iterator.
     */
  readonly function: IFunction;
  constructor(scope: cdk.Construct, name: string, props: IteratorLambdaProps) {
    super(scope, name);
    const iteratorLambdaRole = new iam.Role(this, 'IteratorLambdaRole', {
      assumedBy: new iam.CompositePrincipal(
        new iam.ServicePrincipal('lambda.amazonaws.com'),
      ),
      description: 'An execution role for a Lambda function to invoke a target Lambda Function per time unit less than one minute',
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
        iam.ManagedPolicy.fromAwsManagedPolicyName('AWSXRayDaemonWriteAccess'),
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaVPCAccessExecutionRole'),
      ],
      roleName: 'Lambda-Iterator-Role',
    });
    cdk.DockerVolumeConsistency.CONSISTENT;


    iteratorLambdaRole.addToPolicy(new iam.PolicyStatement({
      sid: 'TargetLambdaPermission',
      effect: iam.Effect.ALLOW,
      actions: ['lambda:InvokeFunction'],
      resources: [props.targetFunction.functionArn],
    }));

    this.function = new NodejsFunction(this, 'Iterator', {
      functionName: 'lambda-subminute-iterator',
      description: 'A function for breaking the limit of 1 minute with the CloudWatch Rules.',
      logRetention: RetentionDays.THREE_MONTHS,
      runtime: Runtime.NODEJS_14_X,
      entry: path.join(__dirname, 'resources/iterator/iterator_agent.ts'),
      handler: 'lambdaHandler',
      environment: {
        TARGET_FN_NAME: props.targetFunction.functionName,
      },
      memorySize: 128,
      role: iteratorLambdaRole,
      timeout: cdk.Duration.seconds(58), // 1 min
      tracing: Tracing.ACTIVE,
    });
  }
}

interface SubminuteStateMachineProps {
  /**
   * the name of the state machine.
   */
  stateMachineName: string;
  /**
   * the Lambda function that executes your intention.
   */
  targetFunction: IFunction;
  /**
   * the iterator Lambda function for the target Lambda function.
   */
  iteratorFunction: IFunction;
}

class SubminuteStateMachine extends cdk.Construct {
  readonly stateMachine: sfn.StateMachine;
  constructor(scope: cdk.Construct, id: string, props: SubminuteStateMachineProps) {
    super(scope, id);
    const stateMachineRole = this._createWorkFlowRole(
      props.targetFunction.functionArn, props.iteratorFunction.functionArn);
    const jobDefinition = this.createJobDefinition(props.iteratorFunction);
    const stateMachine = new sfn.StateMachine(this, 'StateMachine', {
      stateMachineName: props.stateMachineName,
      definition: jobDefinition,
      role: stateMachineRole,
    });
    this.stateMachine = stateMachine;
  }

  /**
   * Creates a state machine for breaking the limit of 1 minute with the CloudWatch Rules.
   *
   * @param iteratorFunction the iterator Lambda function for the target Labmda funciton.
   * @returns THe job definition for the state machine.
   */
  private createJobDefinition = (
    iteratorFunction: IFunction): sfn.Chain => {
    const configureCount = new sfn.Pass(this, 'ConfigureCount', {
      result: sfn.Result.fromObject({
        index: 0,
        count: 6,
      }),
      resultPath: '$.iterator',
    });
    const iterator = new tasks.LambdaInvoke(this, 'Iterator', {
      lambdaFunction: iteratorFunction,
      resultPath: '$.iterator',
      resultSelector: {
        'index.$': '$.Payload.index',
        'count.$': '$.Payload.count',
        'continue.$': '$.Payload.continue',
      },
    });
    const wait = new sfn.Wait(this, 'Wait for the target Lambda function finished', {
      time: sfn.WaitTime.duration(cdk.Duration.seconds(10)),
    });
    wait.next(iterator);
    const done = new sfn.Pass(this, 'Done');

    const isCountReached = new sfn.Choice(this, 'IsCountReached');
    isCountReached.when(sfn.Condition.booleanEquals('$.iterator.continue', true), wait);
    isCountReached.otherwise(done);
    const jobDefinition = configureCount.next(iterator).next(isCountReached);
    return jobDefinition;
  }

  /**
   * Creates a role and corresponding policies for the subminute state machine.
   *
   * @param targetFunctionArn the ARN of the Lambda function that executes your intention.
   * @param iteratorFunctionArn the ARN of the iterator Lambda function for the target Lambda function.
   * @returns the role as the documentation indicates.
   */
  private _createWorkFlowRole = (targetFunctionArn: string, iteratorFunctionArn: string) => {
    const workFlowExecutionRole = new iam.Role(this, 'StepFunctionExecutionRole', {
      assumedBy: new iam.ServicePrincipal('states.amazonaws.com'),
      description: 'Execute a workflow related to executing a Lambda function per time unit less than 1 minute.',
    });
    workFlowExecutionRole.addToPolicy(new iam.PolicyStatement({
      sid: 'LambdaInvokePermissions',
      effect: iam.Effect.ALLOW,
      actions: ['lambda:InvokeFunction'],
      resources: [
        targetFunctionArn,
        iteratorFunctionArn,
      ],
    }));
    return workFlowExecutionRole;
  }
}