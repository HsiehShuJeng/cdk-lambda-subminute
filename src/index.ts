export * from './cdk-lambda-subminute';