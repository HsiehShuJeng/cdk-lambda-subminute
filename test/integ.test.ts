import * as fs from 'fs';
import * as path from 'path';
import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';
import { LambdaSubminute } from '../src';

test('simple test', () => {
  const outdir = typeof process.env.GITHUB_WORKSPACE === 'undefined'? cdk.FileSystem.tmpdir : '/__w/_temp';
  const app = new cdk.App({ outdir: outdir });
  const stack = new cdk.Stack(app, 'TestStack');
  console.log(cdk.AssetStaging.BUNDLING_OUTPUT_DIR);
  console.log(`outdir:${outdir}`);
  console.log(cdk.FileSystem.tmpdir);
  const targetLabmda = new Function(stack, 'targetFunction', {
    code: Code.fromInline('exports.handler = function(event, ctx, cb) { return cb(null, "hi"); })'),
    functionName: 'testTargetFunction',
    runtime: Runtime.NODEJS_12_X,
    handler: 'index.handler',
  });

  new LambdaSubminute(stack, 'LambdaSubminute', { targetFunction: targetLabmda });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(stack).toHaveResourceLike('AWS::Lambda::Function', {
    Role: {
      'Fn::GetAtt': [
        'LambdaSubminuteIteratorLambdaIteratorLambdaRole980819E8',
        'Arn',
      ],
    },
    Environment: {
      Variables: {
        TARGET_FN_NAME: {
          Ref: 'targetFunctionFDF105E2',
        },
      },
    },
    FunctionName: 'lambda-subminute-iterator',
    Handler: 'iterator_agent.handler',
    MemorySize: 128,
    Runtime: 'python3.8',
    Timeout: 58,
    TracingConfig: {
      Mode: 'Active',
    },
  });
});