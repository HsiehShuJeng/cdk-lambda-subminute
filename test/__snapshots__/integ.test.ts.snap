// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple test 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24ArtifactHashBA91B77F": Object {
      "Description": "Artifact hash for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5": Object {
      "Description": "S3 bucket for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861": Object {
      "Description": "S3 key for asset version \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParametersd72315fab02eff25472855563df948e42f1bd998999705cf188cd116e0c44dc2ArtifactHash35D93253": Object {
      "Description": "Artifact hash for asset \\"d72315fab02eff25472855563df948e42f1bd998999705cf188cd116e0c44dc2\\"",
      "Type": "String",
    },
    "AssetParametersd72315fab02eff25472855563df948e42f1bd998999705cf188cd116e0c44dc2S3BucketC26E13A8": Object {
      "Description": "S3 bucket for asset \\"d72315fab02eff25472855563df948e42f1bd998999705cf188cd116e0c44dc2\\"",
      "Type": "String",
    },
    "AssetParametersd72315fab02eff25472855563df948e42f1bd998999705cf188cd116e0c44dc2S3VersionKeyD9CC074E": Object {
      "Description": "S3 key for asset version \\"d72315fab02eff25472855563df948e42f1bd998999705cf188cd116e0c44dc2\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "LambdaSubminuteIteratorLambdaIterator43007E8C": Object {
      "DependsOn": Array [
        "LambdaSubminuteIteratorLambdaIteratorLambdaRoleDefaultPolicy259A62D8",
        "LambdaSubminuteIteratorLambdaIteratorLambdaRole980819E8",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersd72315fab02eff25472855563df948e42f1bd998999705cf188cd116e0c44dc2S3BucketC26E13A8",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd72315fab02eff25472855563df948e42f1bd998999705cf188cd116e0c44dc2S3VersionKeyD9CC074E",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd72315fab02eff25472855563df948e42f1bd998999705cf188cd116e0c44dc2S3VersionKeyD9CC074E",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "A function for breaking the limit of 1 minute with the CloudWatch Rules.",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TARGET_FN_NAME": Object {
              "Ref": "targetFunctionFDF105E2",
            },
          },
        },
        "FunctionName": "lambda-subminute-iterator",
        "Handler": "index.lambdaHandler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaSubminuteIteratorLambdaIteratorLambdaRole980819E8",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 58,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaSubminuteIteratorLambdaIteratorLambdaRole980819E8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "An execution role for a Lambda function to invoke a target Lambda Function per time unit less than one minute",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AWSXRayDaemonWriteAccess",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "Lambda-Iterator-Role",
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaSubminuteIteratorLambdaIteratorLambdaRoleDefaultPolicy259A62D8": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "targetFunctionFDF105E2",
                  "Arn",
                ],
              },
              "Sid": "TargetLambdaPermission",
            },
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaSubminuteIteratorLambdaIteratorLambdaRoleDefaultPolicy259A62D8",
        "Roles": Array [
          Object {
            "Ref": "LambdaSubminuteIteratorLambdaIteratorLambdaRole980819E8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaSubminuteIteratorLambdaIteratorLogRetention5C3C3E8C": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "LambdaSubminuteIteratorLambdaIterator43007E8C",
              },
            ],
          ],
        },
        "RetentionInDays": 90,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LambdaSubminuteStartSubminuteStateMachine54B0F2BF": Object {
      "Properties": Object {
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "A rule to run the subminute state machine, i.e. ",
              Object {
                "Fn::GetAtt": Array [
                  "LambdaSubminuteSubminuteStateMachine23D30D9D",
                  "Name",
                ],
              },
            ],
          ],
        },
        "Name": "subminute-statemachine-lambda-rule",
        "ScheduleExpression": "cron(50/1 15-17 ? * * *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "LambdaSubminuteSubminuteStateMachine23D30D9D",
            },
            "Id": "Target0",
            "Input": "{\\"iterator\\":{\\"index\\":0,\\"count\\":6}}",
            "RoleArn": Object {
              "Fn::GetAtt": Array [
                "LambdaSubminuteSubminuteStateMachineEventsRoleE48C5374",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "LambdaSubminuteSubminuteStateMachine23D30D9D": Object {
      "DependsOn": Array [
        "LambdaSubminuteSubminuteStateMachineStepFunctionExecutionRoleDefaultPolicy93E64C89",
        "LambdaSubminuteSubminuteStateMachineStepFunctionExecutionRoleB6DFA802",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"ConfigureCount\\",\\"States\\":{\\"ConfigureCount\\":{\\"Type\\":\\"Pass\\",\\"Result\\":{\\"index\\":0,\\"count\\":6},\\"ResultPath\\":\\"$.iterator\\",\\"Next\\":\\"Iterator\\"},\\"Iterator\\":{\\"Next\\":\\"IsCountReached\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"ResultPath\\":\\"$.iterator\\",\\"ResultSelector\\":{\\"index.$\\":\\"$.Payload.index\\",\\"count.$\\":\\"$.Payload.count\\",\\"continue.$\\":\\"$.Payload.continue\\"},\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "LambdaSubminuteIteratorLambdaIterator43007E8C",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}},\\"Wait for the target Lambda function finished\\":{\\"Type\\":\\"Wait\\",\\"Seconds\\":10,\\"Next\\":\\"Iterator\\"},\\"IsCountReached\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.iterator.continue\\",\\"BooleanEquals\\":true,\\"Next\\":\\"Wait for the target Lambda function finished\\"}],\\"Default\\":\\"Done\\"},\\"Done\\":{\\"Type\\":\\"Pass\\",\\"End\\":true}}}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "LambdaSubminuteSubminuteStateMachineStepFunctionExecutionRoleB6DFA802",
            "Arn",
          ],
        },
        "StateMachineName": "lambda-subminute-statemachine",
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "LambdaSubminuteSubminuteStateMachineEventsRoleDefaultPolicy88FF3DD8": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "LambdaSubminuteSubminuteStateMachine23D30D9D",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaSubminuteSubminuteStateMachineEventsRoleDefaultPolicy88FF3DD8",
        "Roles": Array [
          Object {
            "Ref": "LambdaSubminuteSubminuteStateMachineEventsRoleE48C5374",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaSubminuteSubminuteStateMachineEventsRoleE48C5374": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaSubminuteSubminuteStateMachineStepFunctionExecutionRoleB6DFA802": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Execute a workflow related to executing a Lambda function per time unit less than 1 minute.",
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaSubminuteSubminuteStateMachineStepFunctionExecutionRoleDefaultPolicy93E64C89": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "targetFunctionFDF105E2",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::GetAtt": Array [
                    "LambdaSubminuteIteratorLambdaIterator43007E8C",
                    "Arn",
                  ],
                },
              ],
              "Sid": "LambdaInvokePermissions",
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "LambdaSubminuteIteratorLambdaIterator43007E8C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaSubminuteSubminuteStateMachineStepFunctionExecutionRoleDefaultPolicy93E64C89",
        "Roles": Array [
          Object {
            "Ref": "LambdaSubminuteSubminuteStateMachineStepFunctionExecutionRoleB6DFA802",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "targetFunctionFDF105E2": Object {
      "DependsOn": Array [
        "targetFunctionServiceRole6EE8FF2F",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "exports.handler = function(event, ctx, cb) { return cb(null, \\"hi\\"); })",
        },
        "FunctionName": "testTargetFunction",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "targetFunctionServiceRole6EE8FF2F",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "targetFunctionServiceRole6EE8FF2F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
